### PR TITLE
Sapservices gatherer 

### PR DESCRIPTION
**Pull request by CDimonaco**
Gatherer for sapservices file, by default `/usr/sap/sapservices`.

Return each entry of sapservices file as a map, with the kind of startup system used, `systemd` or `sapservices`

### CODE REVIEW

File: internal/factsengine/gatherers/sapservices.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+var (
+	SapServicesParsingError = entities.FactGatheringError{
+		Type:    "sap-services-parsing-error",
+		Message: "error parsing the sap services file",
+	}
+	SapServicesFileError = entities.FactGatheringError{
+		Type:    "sap-services-parsing-error",
+		Message: "error reading the sap services file",
+	}
+	SapstartSIDExtractionPattern = regexp.MustCompile(`(?s)pf=([^[:space:]]+)/(.*?)_`)
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:34]: Nothing against this pattern if it works, but there are many ways of getting the SID.
For example, you could get like: `.*pf=/usr/sap/(.*)/.*`. At the end, the sid is the 3rd folder of the `/usr/sap/SID` path

File: internal/factsengine/gatherers/sapservices.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+func extractSIDFromSapstartService(sapServicesContent string) string {
+	matches := SapstartSIDExtractionPattern.FindStringSubmatch(sapServicesContent)
+	if len(matches) != 2 {
+		return ""
+	}
+	return matches[1]
+}
+
+type SapServices struct {
+	fs               afero.Fs
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:69]: Do you need afero if you pass a fixed file path?
For example, in the `groups` gatherer you implemented without it

File: internal/factsengine/gatherers/sapservices.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+		return nil, err
+	}
+
+	var unmarshalled []interface{}
+	err = json.Unmarshal(marshalled, &unmarshalled)
+	if err != nil {
+		return nil, err
+	}
+
+	return entities.NewFactValue(unmarshalled, entities.WithStringConversion())
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:179]: I don't think we need `WithStringConversion`, after all, all the values are strings hehe

File: internal/factsengine/gatherers/sapservices.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+	for fileScanner.Scan() {
+		scannedLine := fileScanner.Text()
+		if strings.HasPrefix(scannedLine, "#") || scannedLine == "" {
+			continue
+		}
+		entry := SapServicesEntry{}
+		entry.Content = scannedLine
+
+		if systemdStartup(scannedLine) {
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:134]: I would most probably implement a `getSystemType` function which returns the `Kind` and based on that, do the parsing, maybe even having a map with the extraction functions like `map[string]func(string) string`, so we reduce a bit the repeated code. But I see that this code is more explicit and easy to read

File: internal/factsengine/gatherers/sapservices.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+	var entries []SapServicesEntry
+
+	for fileScanner.Scan() {
+		scannedLine := fileScanner.Text()
+		if strings.HasPrefix(scannedLine, "#") || scannedLine == "" {
+			continue
+		}
+		entry := SapServicesEntry{}
+		entry.Content = scannedLine
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:132]: Would it make sense to story the `content`, `kind` and `sid` in variables and compose the `SapServicesEntry` at the end? So it is visually easier to see the composition.

**Review by arbulu89 — COMMENTED** [review]
The code logic looks totally correct.
I'm wondering if we should do some integrity check on the line contents. I mean, to check if the line with systemctl or sapstartsrv follows the correct structure

File: internal/factsengine/gatherers/sapservices.go (hunk 3 above)
Comment (CDimonaco) [reply internal/factsengine/gatherers/sapservices.go:179]: you're right

File: internal/factsengine/gatherers/sapservices.go (hunk 4 above)
Comment (CDimonaco) [reply internal/factsengine/gatherers/sapservices.go:134]: I was thinking about that, but I prefer some small repetition, the code is really verbose and can help with readability

File: internal/factsengine/gatherers/sapservices.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+var (
+	SapServicesParsingError = entities.FactGatheringError{
+		Type:    "sap-services-parsing-error",
+		Message: "error parsing the sap services file",
+	}
+	SapServicesFileError = entities.FactGatheringError{
+		Type:    "sap-services-parsing-error",
+		Message: "error reading the sap services file",
+	}
+	SapstartSIDExtractionPattern = regexp.MustCompile(`(?s)pf=([^[:space:]]+)/(.*?)_.*_.*`)
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:34]: Do you group the `space` thing for any reason? I don't think it is used
By the way, wouldn't be the usage of `.+` more appropriate?
I mean `.*` would match even with empty string

**Review by arbulu89 — APPROVED** [review]
Remember to add the gatherer to the list of available gatherers
Maybe you can "mix" the 2 success tests to avoid the "duplicated" warning message

File: internal/factsengine/gatherers/sapservices.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+)
+
+// nolint:gochecknoglobals
+var (
+	SapServicesParsingError = entities.FactGatheringError{
+		Type:    "sap-services-parsing-error",
+		Message: "error parsing the sap services file",
+	}
+	SapServicesFileError = entities.FactGatheringError{
+		Type:    "sap-services-parsing-error",
```
Comment (rtorrero) [inline internal/factsengine/gatherers/sapservices.go:31]: I think you forgot to update this with something descriptive of the error

File: internal/factsengine/gatherers/sapservices_test.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+			Gatherer: "sapservices",
+		},
+	}
+
+	result, err := g.Gather(fr)
+	s.Nil(result)
+	s.EqualError(err, "fact gathering error: sap-services-parsing-error - error reading the sap services file: open /usr/sap/sapservices: file does not exist")
+}
+
+func (s *SapServicesGathererSuite) TestGatheringSIDNotIdentifiedSystemd() {
```
Comment (rtorrero) [inline internal/factsengine/gatherers/sapservices_test.go:37]: The names of the test functions are IMHO not very clear / descriptive

**Review by rtorrero — CHANGES_REQUESTED** [review]
Thanks @CDimonaco, looks good to me overall but the names of the tests and content of the errors need some fixing

File: internal/factsengine/gatherers/sapservices.go (hunk 6 above)
Comment (CDimonaco) [reply internal/factsengine/gatherers/sapservices.go:34]: I made the regexp on regex101 as far as I understood it's not a capture but just a more explicit way to express the things, I don't find also the group in the golang extraction

**Review by rtorrero — APPROVED** [review]
Approving as we had a chat and its now clear what I meant, feel free to fix and merge when ready :+1:

File: internal/factsengine/gatherers/gatherer.go [hunk 9]
```diff
@@ -68,5 +68,8 @@ func StandardGatherers() FactGatherersTree {
 		VerifyPasswordGathererName: map[string]FactGatherer{
 			"v1": NewDefaultPasswordGatherer(),
 		},
+		SapServicesGathererName: map[string]FactGatherer{
```
Comment (arbulu89) [inline internal/factsengine/gatherers/gatherer.go:71]: Alphabetical order please

File: internal/factsengine/gatherers/sapservices.go (hunk 6 above)
Comment (arbulu89) [reply internal/factsengine/gatherers/sapservices.go:34]: Well, in `extractSIDFromSapstartService` you look for a 3 elements size match (which includes, the whole text, 2 groups)

File: internal/factsengine/gatherers/sapservices.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+		var kind SapServicesStartupKind
+		var sid string
+
+		if systemdStartup(scannedLine) {
+			kind = SapServicesSystemdStartup
+			extractedSID := extractSIDFromSystemdService(scannedLine)
+			if extractedSID == "" {
+				return nil, SapServicesParsingError.Wrap(
+					fmt.Sprintf("could not extract sid from systemd sap services entry: %s", scannedLine),
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:140]: I would put `sap` as `SAP` in the text. The same goes for the next error text

File: internal/factsengine/gatherers/sapservices.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+	SapServicesSapstartStartup SapServicesStartupKind = "sapstartsrv"
+	sapServicesDefaultPath                            = "/usr/sap/sapservices"
+	SapServicesGathererName                           = "sapservices"
+)
+
+// nolint:gochecknoglobals
+var (
+	SapServicesParsingError = entities.FactGatheringError{
+		Type:    "sap-services-parsing-error",
+		Message: "error parsing the sap services file",
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapservices.go:28]: I would join the `sap services` to `sapservices`. In the `Type` and in the `Message`.
The same for the next errors

**Review by arbulu89 — APPROVED** [review]
Looks good! Some cosmetic suggestions.
At least fix the alphabetical ordering of the `getherer.go` file

File: internal/factsengine/gatherers/sapservices.go (hunk 11 above)
Comment (CDimonaco) [reply internal/factsengine/gatherers/sapservices.go]: done

File: internal/factsengine/gatherers/sapservices.go (hunk 10 above)
Comment (CDimonaco) [reply internal/factsengine/gatherers/sapservices.go]: done

File: internal/factsengine/gatherers/gatherer.go (hunk 9 above)
Comment (CDimonaco) [reply internal/factsengine/gatherers/gatherer.go]: done

**Pull request merged**